### PR TITLE
Fix index pattern when querying ES and condition when searching logs

### PR DESCRIPTION
**Pull request by belimawr**
## What does this PR do?

Two issues were causing flakiness on integration tests and are fixed by this PR:
1. The pattern used to query ES was not working on serverless, this commit updates it to a pattern that works on both stateful and serverless as well as make it more specific to the indexes/data streams we want to query
2. `findESDocs` did not wait for the data to be indexed, only to a successful query on ES. In some cases the documents the test wanted were not indexed yet, leading to 0 documents being returned and the test failing with no error in the logs/diagnostics. This is fixed by waiting for a document count > 0 and no error.

## Why is it important?

It fixes flaky integration tests.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
~~- [ ] I have made corresponding changes to the documentation~~
~~- [ ] I have made corresponding change to the default configuration files~~
~~- [ ] I have added tests that prove my fix is effective or that my feature works~~
~~- [ ] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)~~
- [x] I have added an integration test or an E2E test

~~## Author's Checklist~~
## How to test this PR locally

Run the integration test that was failing with both serverless and stateful
```
STACK_PROVISIONER=serverless TEST_PLATFORMS="linux/amd64" SNAPSHOT=true mage -v integration:single TestLogIngestionFleetManaged
mage integration:clean
TEST_PLATFORMS="linux/amd64" SNAPSHOT=true mage -v integration:single TestLogIngestionFleetManaged
```

It is needed to clean the deployments because at the moment the integration test framework does not keep track of the type of stack it deployed, more information: https://github.com/elastic/elastic-agent/issues/3756

## Related issues

- Closes https://github.com/elastic/elastic-agent/issues/3741
- Closes https://github.com/elastic/elastic-agent/pull/3743

~~## Use cases~~
~~## Screenshots~~
~~## Logs~~

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

File: pkg/testing/tools/estools/elasticsearch.go [hunk 1]
```diff
@@ -488,7 +488,7 @@ func GetLogsForDatasetWithContext(ctx context.Context, client elastictransport.I
 		},
 	}
 
-	return performQueryForRawQuery(ctx, indexQuery, "*ds-logs*", client)
```
Comment (cmacknz) [inline pkg/testing/tools/estools/elasticsearch.go:491]: This must have been querying the underlying data stream backing indexes directly then? That wasn't correct if so.

Everything follows the data stream naming scheme so `log*` is correct if you are querying for all logs data streams.

In our own tests for elastic agent logs you should be able to do `logs-elastic_agent*` but this ES instance contains such few documents I'm not sure this make much of a difference.

File: testing/integration/logs_ingestion_test.go [hunk 2]
```diff
@@ -104,15 +104,11 @@ func testMonitoringLogsAreShipped(
 ) {
 	// Stage 1: Make sure metricbeat logs are populated
 	t.Log("Making sure metricbeat logs are populated")
-	require.Eventually(t,
```
Comment (cmacknz) [inline testing/integration/logs_ingestion_test.go:107]: Will removing the eventually here make this flaky again?

File: testing/integration/logs_ingestion_test.go [hunk 3]
```diff
@@ -199,14 +195,21 @@ func testMonitoringLogsAreShipped(
 
 func findESDocs(t *testing.T, findFn func() (estools.Documents, error)) estools.Documents {
 	var docs estools.Documents
+	start := time.Now()
+	defer func() {
+		fmt.Println(">>>>>>>>>>>>>>>>>>>> Took", time.Now().Sub(start))
```
Comment (cmacknz) [inline testing/integration/logs_ingestion_test.go:200]:
```suggestion
		t.Log(">>>>>>>>>>>>>>>>>>>> Took", time.Now().Sub(start))
```

Same applies for the one below, assuming you plan to keep these and they aren't just for debugging

File: testing/integration/logs_ingestion_test.go [hunk 4]
```diff
@@ -199,14 +195,21 @@ func testMonitoringLogsAreShipped(
...
+	}()
+	count := 0
 	require.Eventually(
 		t,
 		func() bool {
+			count++
+			fmt.Println(">>>>>>>>>>>>>>>>>>>> Iteration ", count)
```
Comment (cmacknz) [inline testing/integration/logs_ingestion_test.go:207]:
```suggestion
			t.Logf(">>>>>>>>>>>>>>>>>>>> Iteration %d", count)
```

File: testing/integration/logs_ingestion_test.go (hunk 2 above)
Comment (cmacknz) [reply testing/integration/logs_ingestion_test.go:107]: Ah I see there is a require eventually inside findESDocs.

File: pkg/testing/tools/estools/elasticsearch.go (hunk 1 above)
Comment (cmacknz) [reply pkg/testing/tools/estools/elasticsearch.go:491]: We do this in other places in this file 

https://github.com/elastic/elastic-agent/blob/f7dcbd74ce0e27d3b46b478755168c66d78dc111/pkg/testing/tools/estools/elasticsearch.go#L437

https://github.com/elastic/elastic-agent/blob/f7dcbd74ce0e27d3b46b478755168c66d78dc111/pkg/testing/tools/estools/elasticsearch.go#L365

Comment (cmacknz) [reply pkg/testing/tools/estools/elasticsearch.go:491]: I also see this filters by dataset already which should catch bugs where the data exists but in the wrong index, but limiting the query only to the index things are supposed to be in would be a stricter test.

File: testing/integration/logs_ingestion_test.go (hunk 2 above)
Comment (belimawr) [reply testing/integration/logs_ingestion_test.go:107]: Exactly.

File: testing/integration/logs_ingestion_test.go (hunk 3 above)
Comment (belimawr) [reply testing/integration/logs_ingestion_test.go:200]: That's just for debugging, I'm still trying to understand why it only fails to find the docs in CI....

**Comment by belimawr** [comment]
The failure seems to be caused by: https://github.com/elastic/elastic-agent/issues/3664

**Comment by belimawr** [comment]
buildkite test this

File: testing/integration/endpoint_security_test.go [hunk 5]
```diff
@@ -493,7 +493,7 @@ func TestEndpointSecurityNonDefaultBasePath(t *testing.T) {
 	info := define.Require(t, define.Requirements{
 		Stack:   &define.Stack{},
 		Local:   false, // requires Agent installation
-		Isolate: false,
+		Isolate: true,
```
Comment (blakerouse) [inline testing/integration/endpoint_security_test.go:496]: Why all the changes to isolate? That means this will add 4 more machines to perform tests, why?

This seems unrelated to the change implemented in this PR. I didn't see anything in the PR description describing this change and its reason.

**Review by blakerouse — CHANGES_REQUESTED** [review]

File: testing/integration/endpoint_security_test.go (hunk 5 above)
Comment (belimawr) [reply testing/integration/endpoint_security_test.go]: That was me trying to get the integration tests to pass. Some exploratory testing because I just can't reproduce any of the CI failures when running the tests from my machine. I undid this change.

But I still have the changes from https://github.com/elastic/elastic-agent/pull/3786 to try getting CI happy. 

I'll request a re-review once it's the final code that can pass CI.

File: pkg/testing/tools/estools/elasticsearch.go (hunk 1 above)
Comment (belimawr) [reply pkg/testing/tools/estools/elasticsearch.go:491]: Thanks Craig! I'll update it to be more strict then.

**Comment by belimawr** [comment]
Folks, this PR is finally (well, hopefully) ready for a final review. I found another source of flakiness and fixed it in my last commit ff4edf2824.

**Review by cmacknz — APPROVED** [review]

**Review by blakerouse — APPROVED** [review]
Looks good!
